toolchain/lower/file_context.cpp

```diff
@@ -397,25 +397,45 @@ auto FileContext::BuildFunctionDefinition(SemIR::FunctionId function_id)
                         .GetAs<SemIR::FunctionDecl>(function.latest_decl_id())
                         .decl_block_id;
   }
-  llvm::ArrayRef<SemIR::InstBlockId> singleton_decl_block = decl_block_id;
 
-  // Lower all blocks.
-  for (auto block_id : llvm::concat<const SemIR::InstBlockId>(
-           singleton_decl_block, body_block_ids)) {
+  // Lowers the contents of block_id into the corresponding LLVM block,
+  // creating it if it doesn't already exist.
+  auto lower_block = [&](SemIR::InstBlockId block_id) {
     CARBON_VLOG("Lowering {0}\n", block_id);
     auto* llvm_block = function_lowering.GetBlock(block_id);
     // Keep the LLVM blocks in lexical order.
     llvm_block->moveBefore(llvm_function->end());
     function_lowering.builder().SetInsertPoint(llvm_block);
-    function_lowering.LowerBlock(block_id);
+    function_lowering.LowerBlockContents(block_id);
+    return llvm_block;
+  };
+
+  auto* llvm_decl_block = lower_block(decl_block_id);
+
+  // If the decl block is empty, reuse it as the first body block. We don't do
+  // this when the decl block is non-empty so that any branches back to the
+  // first body block don't also re-execute the decl.
+  if (llvm_decl_block->empty()) {
+    CARBON_CHECK(function_lowering.TryToReuseBlock(body_block_ids.front(),
+                                                   llvm_decl_block));
+  } else {
+    function_lowering.builder().SetInsertPoint(llvm_decl_block);
+    function_lowering.builder().CreateBr(
+        function_lowering.GetBlock(body_block_ids.front()));
```

Comment:
Maybe
```suggestion
  // If the decl block is empty, reuse it as the first body block. We don't do
  // this when the decl block is non-empty so that any branches back to the
  // first body block don't also re-execute the decl.
  llvm::BasicBlock* block = context.builder().GetInsertBlock();
  if (block->empty() &&
      function_lowering.TryToReuseBlock(body_block_ids.front(),
                                        block))) {
    // Reuse this block as the first block of the function body.
  } else {
    function_lowering.builder().CreateBr(
        function_lowering.GetBlock(body_block_ids.front()));
```
to make this more similar to the other uses of `TryToReuseBlock` and handle the possibility that we somehow generate multiple LLVM basic blocks while lowering the decl block. (We shouldn't need to `SetInsertionPoint` because it should still be in the `llvm_decl_block`.) I wonder if we should actually factor out a function to do this block-reuse or branch process -- we seem to do the same thing [three](https://github.com/carbon-language/carbon-lang/blob/954441c358d1b7df5515ab2728953bce0eb690d4/toolchain/lower/handle.cpp#L104) [times](https://github.com/carbon-language/carbon-lang/blob/954441c358d1b7df5515ab2728953bce0eb690d4/toolchain/lower/handle.cpp#L135) now.

Reply:
Done, with some tweaks. I assume removing the first `lower_block` call was an accident?

Reply:
Yes, oops, I just meant to remove the variable not the initializer.